requestmanager/executor/executor.go

```diff
@@ -131,23 +128,20 @@ func (e *Executor) traverse(rt RequestTask) error {
 		lnk, linkContext := rt.Traverser.CurrentRequest()
 		// attempt to load
 		log.Debugf("will load link=%s", lnk)
-		resultChan := e.loader(rt.P, rt.Request.ID(), lnk, linkContext)
-		var result types.AsyncLoadResult
-		// check for immediate result
-		select {
-		case result = <-resultChan:
-		default:
-			// if no immediate result
-			// initiate remote request if not already sent (we want to fill out the doNotSendCids on a resume)
-			if err := onlyOnce.startRemoteRequest(); err != nil {
+		result := rt.ReconciledLoader.BlockReadOpener(linkContext, lnk)
+		// if we've only loaded locally so far and hit a missing block
+		// initiate remote request and retry the load operation from remote
+		if _, ok := result.Err.(graphsync.MissingBlockErr); ok && !requestSent {
+			requestSent = true
```

Comment:
Should this be a state that's available on the loader rather than having to figure it out here? It seems a little risky having to make inferences like this from the outside. It seems to me that there's a new loader made for each request in the requestmanager, and then this assumption is implicit in the executor.

It's also a bit weird that we're asking the loader for a block, and then using the result that the loader gives us to then tell the loader what its own state is. It seems that we're doing state management here so we can trigger the `RetryLastOfflineLoad()`. But the flow does raise questions about who's responsible for what. The loader clearly isn't just a dumb container, yet it's not responsible for its own state management.

Reply:
yea, I mean, maybe? honestly this is a boundary I thought about a lot and couldn't decide upon.

Honestly, I thought about having the loader just make the request itself cause yea, then state in only one place... I dunno, at that point it's just doing a whole whole lot and it makes this already large PR even bigger. 

I'm not sure where to draw the line, and my gut feeling is: let's ship it, prove it works, and then refactor more as needed.

Reply:
fair enough, I'll buy that